Spigot-API-Patches/0132-Custom-plugin-advancement-API.patch

```diff
@@ -0,0 +1,260 @@
++            this.title = title;
++            this.description = description;
++            this.icon = icon;
++            this.background = background;
++            this.frameType = frameType;
++            this.showToast = showToast;
++            this.announceToChat = announceToChat;
++            this.hidden = hidden;
++        }
++
++        public String getTitle() {
++            return title;
++        }
++
++        public String getDescription() {
++            return description;
++        }
++
++        public ItemStack getIcon() {
++            return icon;
++        }
++
++        @Nullable
++        public NamespacedKey getBackground() {
++            return background;
++        }
++
++        public DisplayFrameType getFrameType() {
++            return frameType;
++        }
++
++        public boolean showToast() {
++            return showToast;
++        }
++
++        public boolean announceToChat() {
++            return announceToChat;
++        }
++
++        public boolean isHidden() {
++            return hidden;
++        }
++    }
++
++    public enum DisplayFrameType {
++        TASK, CHALLENGE, GOAL
++    }
++}
+diff --git a/src/main/java/org/bukkit/advancement/AdvancementManager.java b/src/main/java/org/bukkit/advancement/AdvancementManager.java
+new file mode 100644
+index 00000000..1bf5611d
+--- /dev/null
++++ b/src/main/java/org/bukkit/advancement/AdvancementManager.java
+@@ -0,0 +1,24 @@
++package org.bukkit.advancement;
++
++import org.bukkit.plugin.Plugin;
++
++import java.util.List;
++
++/**
++ * Manager of Advancements
++ */
++public interface AdvancementManager {
++    /**
++     * Clear a plugin's custom advancements
++     * @param plugin the plugin to clear the advancements of
++     */
++    void clearAdvancements(Plugin plugin);
++
++    /**
++     * Registers an advancement to the plugin.
++     * Note: Changes do not take effect until the next tick or Bukkit.reloadData() is called
++     * @param plugin the plugin to register the advancement to
++     * @param advancementData the advancement data to register
++     */
++    void registerAdvancement(Plugin plugin, AdvancementData advancementData);
++}
+diff --git a/src/main/java/org/bukkit/plugin/SimplePluginManager.java b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+index cb2b0b9c..8dfacb04 100644
+--- a/src/main/java/org/bukkit/plugin/SimplePluginManager.java
++++ b/src/main/java/org/bukkit/plugin/SimplePluginManager.java
+@@ -468,6 +468,8 @@ public final class SimplePluginManager implements PluginManager {
+                 handlePluginException("Error occurred (in the plugin loader) while unregistering plugin channels for "
+                         + plugin.getDescription().getFullName() + " (Is it up to date?)", ex, plugin); // Paper
+             }
++
++            server.getAdvancementManager().clearAdvancements(plugin); // Paper - custom advancements API
```

Comment:
How does this affect saving?

Reply:
in my experience advancements save even if they are invalid, so if the advancements disappear then they stay in the player data so if you add them back they still work

Reply:
Needs testing

Reply:
(my advancements break on a daily basis lol) but yes i will test it

Reply:
OK it seems I actually made a workaround for this a very long time ago, where I manually add it back. Advancements are only loaded if they exist and they only save if they're loaded, so it would break whenever a player's data was saved from what I can tell. What would be the best workaround? I was thinking maybe store the advancements that failed to load and aggregate them when saving?